apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts

```diff
@@ -87,14 +144,100 @@ Keep in mind you are not just generating an example; you are generating an actua
+ * @param originalPrompt The prompt creating the original generation.
+ * @param revisionPrompt A prompt with user instructions on how to revise the question.
+ * @param originalHTML The question.html file to revise.
+ * @param originalPython The server.py file to revise.
+ * @returns A server job ID for the generation task and a promise to return the associated saved data on completion.
+ */
+export async function regenerateQuestion(
+  client: OpenAI,
+  courseId: string | undefined,
+  authnUserId: string,
+  originalPrompt: string,
+  revisionPrompt: string,
+  originalHTML: string,
+  originalPython: string,
+): Promise<{ jobSequenceId: string; jobData: ServerJobResult }> {
+  const serverJob = await createServerJob({
+    courseId,
+    type: 'llm_question_regen',
+    description: 'Revise a question using the LLM',
+  });
+
+  const jobPromise = serverJob.execute(async (job) => {
+    job.info(`prompt is ${revisionPrompt}`);
+
+    const context = await makeContext(client, originalPrompt, authnUserId);
+    const sysPrompt = `
+        ${promptPreamble(context)}
+        # Previous Generations
+
+        A user previously used the assistant to generate a question with following prompt:
+
+        ${originalPrompt}
+
+        You generated the following:
+
+       ${
+         originalHTML === undefined
+           ? ''
+           : `\`\`\`html
+        ${originalHTML}
+        \`\`\``
+       }
+
+      ${
+        originalPython === undefined
+          ? ''
+          : `\`\`\`python
+        ${originalPython}
+        \`\`\``
+      }
+        
+        # Prompt
+        
+        A user will now request your help in in revising the question that you generated. Respond in a friendly but concise way. Include \`question.html\` and \`server.py\` in Markdown code fences in your response, and tag each code fence with the language (either \`html\` or \`python\`). Omit \`server.py\` if the question does not require it (for instance, if the question does not require randomization).
+        
+        Keep in mind you are not just generating an example; you are generating an actual question that the user will use directly.
+        `;
+
+    job.info(`system prompt is: ${sysPrompt}`);
+
+    //TODO [very important]: normalize to prevent prompt injection attacks
+
+    const completion = await client.chat.completions.create({
+      model: 'gpt-3.5-turbo',
+      messages: [
+        { role: 'system', content: sysPrompt },
+        { role: 'user', content: revisionPrompt },
+      ],
+      user: openAiUserFromAuthn(authnUserId),
+    });
 
-    // TODO: Process generated documents
+    extractFromCompletion(completion, job);
+    job.data['prompt'] = revisionPrompt;
+    job.data['originalPrompt'] = originalPrompt;
+    job.data['generation'] = completion.choices[0].message.content;
+    job.data['context'] = context;
   });
 
-  return serverJob.jobSequenceId;
+  const jobData = await jobPromise;
```

Comment:
It's a bit clunky to store the promise and then await it later. Instead you can remove this line and above just do:
```
const jobData = await serverJob.execute(async (job) => {
```

Reply:
Changed in b5c604cd1a9f33f625b695563a12dcab48f3d0f9.